tempodb/tempodb.go

```diff
@@ -309,77 +310,102 @@ func (rw *readerWriter) pollBlocklist() {
 	rw.cleanMissingTenants(tenants)
 
 	for _, tenantID := range tenants {
-		blockIDs, err := rw.r.Blocks(ctx, tenantID)
+
+		newBlockList, newCompactedBlockList, err := rw.pollTenant(ctx, tenantID)
+
 		if err != nil {
 			metricBlocklistErrors.WithLabelValues(tenantID).Inc()
-			level.Error(rw.logger).Log("msg", "error polling blocklist", "tenantID", tenantID, "err", err)
+			level.Error(rw.logger).Log("msg", "run blocklist jobs", "tenantID", tenantID, "err", err)
+			continue
 		}
 
-		interfaceSlice := make([]interface{}, 0, len(blockIDs))
-		for _, id := range blockIDs {
-			interfaceSlice = append(interfaceSlice, id)
-		}
+		metricBlocklistLength.WithLabelValues(tenantID).Set(float64(len(newBlockList)))
 
-		listMutex := sync.Mutex{}
-		blocklist := make([]*backend.BlockMeta, 0, len(blockIDs))
-		compactedBlocklist := make([]*backend.CompactedBlockMeta, 0, len(blockIDs))
-		_, err = rw.pool.RunJobs(ctx, interfaceSlice, func(ctx context.Context, payload interface{}) ([]byte, error) {
-			blockID := payload.(uuid.UUID)
-
-			var compactedBlockMeta *backend.CompactedBlockMeta
-			blockMeta, err := rw.r.BlockMeta(ctx, blockID, tenantID)
-			// if the normal meta doesn't exist maybe it's compacted.
-			if err == backend.ErrMetaDoesNotExist {
-				blockMeta = nil
-				compactedBlockMeta, err = rw.c.CompactedBlockMeta(blockID, tenantID)
-			}
+		rw.blockListsMtx.Lock()
+		rw.blockLists[tenantID] = newBlockList
+		rw.compactedBlockLists[tenantID] = newCompactedBlockList
+		rw.blockListsMtx.Unlock()
+	}
+}
 
-			// blocks in intermediate states may not have a compacted or normal block meta.
-			//   this is not necessarily an error, just bail out
-			if err == backend.ErrMetaDoesNotExist {
-				return nil, nil
-			}
+func (rw *readerWriter) pollTenant(ctx context.Context, tenantID string) ([]*backend.BlockMeta, []*backend.CompactedBlockMeta, error) {
+	blockIDs, err := rw.r.Blocks(ctx, tenantID)
+	if err != nil {
+		metricBlocklistErrors.WithLabelValues(tenantID).Inc()
+		level.Error(rw.logger).Log("msg", "error polling blocklist", "tenantID", tenantID, "err", err)
+		return nil, nil, err
+	}
 
-			if err != nil {
-				metricBlocklistErrors.WithLabelValues(tenantID).Inc()
-				level.Error(rw.logger).Log("msg", "failed to retrieve block meta", "tenantID", tenantID, "blockID", blockID, "err", err)
-				return nil, nil
+	bg := tempoUtil.NewBoundedWaitGroup(1)
```

Comment:
Per your comment, it would be nice to make this configurable, and default to 50?